vendors/swedbank-pay-core-library/src/SwedbankPay/Core/Library/Methods/Invoice.php

```diff
@@ -0,0 +1,374 @@
+                break;
+            case 'Failed':
+                $message = isset($transaction['failedReason']) ? $transaction['failedReason'] : 'Capture is failed.';
+                throw new Exception($message);
+            default:
+                throw new Exception('Capture is failed.');
+        }
+
+        return $result;
+    }
+
+    /**
+     * Cancel Invoice.
+     *
+     * @param mixed $orderId
+     * @param int|float|null $amount
+     * @param int|float $vatAmount
+     *
+     * @return Response
+     * @throws Exception
+     */
+    public function cancelInvoice($orderId, $amount = null, $vatAmount = 0)
+    {
+        /** @var Order $order */
+        $order = $this->getOrder($orderId);
+
+        if (!$amount) {
+            $amount = $order->getAmount();
+            $vatAmount = $order->getVatAmount();
+        }
+
+        if (!$this->canCancel($orderId, $amount)) {
+            throw new Exception('Cancellation is not available.');
+        }
+
+        $paymentId = $order->getPaymentId();
+        if (empty($paymentId)) {
+            throw new Exception('Unable to get payment ID');
+        }
+
+        $href = $this->fetchPaymentInfo($paymentId)->getOperationByRel('create-cancellation');
+        if (empty($href)) {
+            throw new Exception('Cancellation is unavailable');
+        }
+
+        $params = [
+            'transaction' => [
+                'activity' => 'FinancingConsumer',
+                'description' => sprintf('Cancellation for Order #%s', $order->getOrderId()),
+                'payeeReference' => $this->generatePayeeReference($orderId)
+            ],
+        ];
+
+        $result = $this->request('POST', $href, $params);
+
+        // Save transaction
+        $transaction = $result['cancellation']['transaction'];
+        $this->saveTransaction($orderId, $transaction);
+
+        switch ($transaction['state']) {
+            case 'Completed':
+                $this->updateOrderStatus(OrderInterface::STATUS_CANCELLED, 'Transaction is cancelled.');
+                break;
+            case 'Initialized':
+            case 'AwaitingActivity':
+                $this->updateOrderStatus(OrderInterface::STATUS_CANCELLED,
+                    sprintf('Transaction cancellation status: %s.', $transaction['state']));
+                break;
+            case 'Failed':
+                $message = isset($transaction['failedReason']) ? $transaction['failedReason'] : 'Cancellation is failed.';
+                throw new Exception($message);
+            default:
+                throw new Exception('Capture is failed.');
+        }
+
+        return $result;
+    }
+
+    /**
+     * Refund Invoice.
+     *
+     * @param mixed $orderId
+     * @param int|float|null $amount
+     * @param int|float $vatAmount
+     *
+     * @return Response
+     * @throws Exception
+     */
+    public function refundInvoice($orderId, $amount = null, $vatAmount = 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method refundInvoice() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-woocommerce-payments/pullRequest?prid=5340032)